libs/prelude/Prelude/Interfaces.idr

```diff
@@ -372,6 +372,145 @@ namespace Foldable
+||| Common examples are `Either` and `Pair`.
+public export
+interface Bifoldable p where
+  bifoldr : (a -> acc -> acc) -> (b -> acc -> acc) -> acc -> p a b -> acc
+
+  bifoldl : (acc -> a -> acc) -> (acc -> b -> acc) -> acc -> p a b -> acc
+  bifoldl f g z t = bifoldr (flip (.) . flip f) (flip (.) . flip g) id t z
+
+  binull : p a a -> Lazy Bool
+  binull = bifoldr {acc = Lazy Bool} (\ _,_ => False) (\ _,_ => False) True
+
+||| Left associative monadic bifold over a structure.
+public export
+bifoldlM :  (Bifoldable p, Monad m)
+         => (f: a -> b -> m a)
+         -> (g: a -> c -> m a)
+         -> (init: a)
+         -> (input: p b c) -> m a
+bifoldlM f g a0 = bifoldl (\ma,b => ma >>= flip f b)
+                          (\ma,c => ma >>= flip g c)
+                          (pure a0)
+
+||| Combines the elements of a structure,
+||| given ways of mapping them to a common monoid.
+public export
+bifoldMap : (Bifoldable p, Monoid m) => (a -> m) -> (b -> m) -> p a b -> m
+bifoldMap f g = bifoldr ((<+>) . f) ((<+>) . g) neutral
+
+||| Combines the elements of a structure using a monoid.
+public export
+biconcat : (Bifoldable p, Monoid m) => p m m -> m
+biconcat = bifoldr (<+>) (<+>) neutral
+
+||| Combines the elements of a structure,
+||| given ways of mapping them to a common monoid.
+public export
+biconcatMap : (Bifoldable p, Monoid m) => (a -> m) -> (b -> m) -> p a b -> m
+biconcatMap f g = bifoldr ((<+>) . f) ((<+>) . g) neutral
+
+||| The conjunction of all elements of a structure containing lazy boolean
+||| values.  `biand` short-circuits from left to right, evaluating until either an
+||| element is `False` or no elements remain.
+public export
+biand : Bifoldable p => p (Lazy Bool) (Lazy Bool) -> Bool
+biand = bifoldl (&&) (&&) True
+
+||| The disjunction of all elements of a structure containing lazy boolean
+||| values.  `bior` short-circuits from left to right, evaluating either until an
+||| element is `True` or no elements remain.
+public export
+bior : Bifoldable p => p (Lazy Bool) (Lazy Bool) -> Bool
+bior = bifoldl (||) (||) False
+
+||| The disjunction of the collective results of applying a predicate to all
+||| elements of a structure. `biany` short-circuits from left to right.
+public export
+biany : Bifoldable p => (a -> Bool) -> (b -> Bool) -> p a b -> Bool
+biany f g = bifoldl (\x,y => x || f y) (\x,y => x || g y) False
+
+||| The disjunction of the collective results of applying a predicate to all
+||| elements of a structure.  `biall` short-circuits from left to right.
+public export
+biall : Bifoldable p => (a -> Bool) -> (b -> Bool) -> p a b -> Bool
+biall f g = bifoldl (\x,y => x && f y) (\x,y => x && g y) True
+
+||| Add together all the elements of a structure.
+public export
+bisum : (Bifoldable p, Num a) => p a a -> a
+bisum = bifoldr (+) (+) 0
+
+||| Add together all the elements of a structure.
+||| Same as `bisum` but tail recursive.
+export
```

Comment:
I'm not convinced all of these should go in the prelude.
Remember that the criterion is: needed for almost any program.
This should probably go into `Data.Bifoldable` in `base`.

Reply:
Moved most utility functions to Data.Bifoldable in base.